prow/cache/cache.go

```diff
@@ -275,3 +277,56 @@ func (lruCache *LRUCache) GetOrAdd(
 
 	return promise.val, ok, promise.err
 }
+
+// AddConstructed adds a new key and value into the LRUCache, but with the
+// understanding that the value is already constructed. Because we don't need to
+// worry about constructing it, the valConstructor and valConstructionPending
+// fields are left uninitialized.
+func (lruCache *LRUCache) AddConstructed(
+	key interface{},
+	val interface{}) {
+
+	lruCache.Lock()
+	promise := &Promise{
+		val:         val,
+		err:         nil,
+		constructed: true,
+	}
+	_ = lruCache.Add(key, promise)
+	lruCache.Unlock()
+}
+
+// GetConstructedEntries returns a copy of the cache but with only the keys and
+// values that are fully constructed (and whose construction did not result in
+// an error). Note that the values here are not Promise types because it would
+// be redundant (we only look at resolved promises).
+func (lruCache *LRUCache) GetConstructedEntries() map[interface{}]interface{} {
+
+	constructed := make(map[interface{}]interface{})
+
+	lruCache.Lock()
```

Comment:
could this be a read lock?

Reply:
Indeed. I changed the underlying mutex to a RWMutex instead of just Mutex.